django/template/defaultfilters.py

```
@@ -517,11 +517,11 @@ def first(value):
 @register.filter(is_safe=True, needs_autoescape=True)
 def join(value, arg, autoescape=True):
     """Join a list with a string, like Python's ``str.join(list)``."""
-    if autoescape:
-        value = [conditional_escape(v) for v in value]
     try:
+        if autoescape:
+            value = [conditional_escape(v) for v in value]
         data = conditional_escape(arg).join(value)
-    except AttributeError:  # fail silently but nicely
+    except TypeError:  # fail silently but nicely
```

I'd update the comment to say `# Fail silently if arg isn't iterable.`